### PR TITLE
add double caching for yahoo to speed up extracted reading.

### PR DESCRIPTION
**Pull request by erip**
Reference pytorch/text#1526

### CODE REVIEW

File: torchtext/datasets/yahooanswers.py [hunk 1]
```diff
@@ -42,13 +42,27 @@ def YahooAnswers(root: str, split: Union[Tuple[str], str]):
 
     cache_dp = url_dp.on_disk_cache(
         filepath_fn=lambda x: os.path.join(root, _PATH),
-        hash_dict={os.path.join(root, _PATH): MD5}, hash_type="md5"
+        hash_dict={os.path.join(root, _PATH): MD5},
+        hash_type="md5"
     )
     cache_dp = GDriveReader(cache_dp).end_caching(mode="wb", same_filepath_fn=True)
     cache_dp = FileOpener(cache_dp, mode="b")
```
Comment (Nayef211) [inline torchtext/datasets/yahooanswers.py:49]: nit: should we rename to `cache_compressed_dp` to be consistent with the [AmazonReviewPolarity PR](https://github.com/pytorch/text/pull/1527/files#diff-daeb7a9b451f1cb20240d2fda55074b6f03d48849646c931c513050af8eff1a5R43-R46)?

File: torchtext/datasets/yahooanswers.py [hunk 2]
```diff
@@ -42,13 +42,27 @@ def YahooAnswers(root: str, split: Union[Tuple[str], str]):
...
+        return file_path
 
-    filter_extracted_files = extracted_files.filter(lambda x: _EXTRACTED_FILES[split] in x[0])
+    cache_dp = cache_dp.on_disk_cache(
+        filepath_fn=extracted_filepath_fn
+    )
+    cache_dp = cache_dp.read_from_tar()
+
+    cache_dp = cache_dp.filter(lambda x: _EXTRACTED_FILES[split] in x[0])
+    cache_dp = cache_dp.end_caching(mode="wb", same_filepath_fn=True)
```
Comment (Nayef211) [inline torchtext/datasets/yahooanswers.py:64]: nit: should we rename `cache_dp` to `cache_decompressed_dp` to be consistent with the AmazonReviewPolarity PR?

File: torchtext/datasets/yahooanswers.py [hunk 3]
```diff
@@ -42,13 +42,27 @@ def YahooAnswers(root: str, split: Union[Tuple[str], str]):
...
+
+    cache_dp = FileOpener(cache_dp, mode="b")
```
Comment (Nayef211) [inline torchtext/datasets/yahooanswers.py:66]: nit: let's call this `data_dp` for consistency with AmazonReviewPolarity PR

**Review by Nayef211 — APPROVED** [review]
This looks great thanks for making the changes!

File: torchtext/datasets/yahooanswers.py [hunk 4]
```diff
@@ -33,22 +33,28 @@
...
+    cache_compressed_dp = url_dp.on_disk_cache(
         filepath_fn=lambda x: os.path.join(root, _PATH),
-        hash_dict={os.path.join(root, _PATH): MD5}, hash_type="md5"
+        hash_dict={os.path.join(root, _PATH): MD5},
+        hash_type="md5"
     )
-    cache_dp = GDriveReader(cache_dp).end_caching(mode="wb", same_filepath_fn=True)
-    cache_dp = FileOpener(cache_dp, mode="b")
+    cache_compressed_dp = GDriveReader(cache_compressed_dp).end_caching(mode="wb", same_filepath_fn=True)
+    cache_compressed_dp = FileOpener(cache_compressed_dp, mode="b")
```
Comment (parmeet) [inline torchtext/datasets/yahooanswers.py:49]: same suggestion as here (https://github.com/pytorch/text/pull/1530#discussion_r788931995)

Comment (erip) [reply torchtext/datasets/yahooanswers.py:49]: Done!

**Review by parmeet — COMMENTED** [review]
LGTM!

**Review by parmeet — APPROVED** [review]

**Pull request merged**